File: java/src/test/java/kanzi/test/TestFunctions.java
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package kanzi.test;

import java.util.Arrays;
import java.util.Random;
import kanzi.ByteFunction;
import kanzi.SliceByteArray;
import kanzi.function.LZ4Codec;
import kanzi.function.RLT;
import kanzi.function.ROLZCodec;
import kanzi.function.SnappyCodec;
import kanzi.function.ZRLT;
import org.junit.Assert;
import org.junit.Test;


public class TestFunctions
{
   public static void main(String[] args)
   {
      if (args.length == 0)
      {
          args = new String[] { "-TYPE=ALL" };
      }

      String type = args[0].toUpperCase();

      if (type.startsWith("-TYPE=")) 
      {
          type = type.substring(6);
          System.out.println("Transform: " + type);

          if (type.equals("ALL"))
          {
             System.out.println("\n\nTestLZ4");
             
             if (testCorrectness("LZ4") == false)
                System.exit(1);
             
             testSpeed("LZ4");
             System.out.println("\n\nTestROLZ");
             
             if (testCorrectness("ROLZ") == false)
                System.exit(1);
             
             testSpeed("ROLZ"); 
             System.out.println("\n\nTestSnappy");
             
             if (testCorrectness("SNAPPY") == false)
                System.exit(1);
             
             testSpeed("SNAPPY");
             System.out.println("\n\nTestZRLT");
             
             if (testCorrectness("ZRLT") == false)
                System.exit(1);
             
             testSpeed("ZRLT");
             System.out.println("\n\nTestRLT");
             
             if (testCorrectness("RLT") == false)
                System.exit(1);
             
             testSpeed("RLT");                 
           }
          else
          {
             System.out.println("Test" + type);
             
             if (testCorrectness(type) == false)
                System.exit(1);
             
             testSpeed(type);
          }
      }        
   }

    
   @Test
   public void testFunctions()
   {
      System.out.println("\n\nTestLZ4");
      Assert.assertTrue(testCorrectness("LZ4"));
      //testSpeed("LZ4");
      System.out.println("\n\nTestROLZ");
      Assert.assertTrue(testCorrectness("ROLZ"));
      //testSpeed("ROLZ");   
      System.out.println("\n\nTestSnappy");
      Assert.assertTrue(testCorrectness("SNAPPY"));
      //testSpeed("SNAPPY");
      System.out.println("\n\nTestZRLT");
      Assert.assertTrue(testCorrectness("ZRLT"));
      //testSpeed("ZRLT");
      System.out.println("\n\nTestRLT");
      Assert.assertTrue(testCorrectness("RLT"));
      //testSpeed("RLT");   
   }
   
   
   private static ByteFunction getByteFunction(String name)
   {
      switch(name) 
      {
         case "LZ4":
            return new LZ4Codec();

         case "SNAPPY":
            return new SnappyCodec();

         case "ZRLT":
            return new ZRLT();

         case "RLT":
            return new RLT();

         case "ROLZ":
            return new ROLZCodec();

         default:
            System.out.println("No such byte function: "+name);
            return null;
      }
   }

    
   private static boolean testCorrectness(String name)
   {        
      byte[] input;
      byte[] output;
      byte[] reverse;
      Random rnd = new Random();

      // Test behavior
      System.out.println("Correctness test for " + name);
      int range = name.equals("ZRLT") ? 5 : 256;

      for (int ii=0; ii<20; ii++)
      {
         System.out.println("\nTest "+ii);
         int[] arr;

         if (ii == 0)
         {
            arr = new int[] {
               0, 1, 2, 2, 2, 2, 7, 9,  9, 16, 16, 16, 1, 3,
              3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3
            };
         }
         else if (ii == 1)
         {
            arr = new int[66000];
            arr[0] = 1;

            for (int i=1; i<arr.length; i++)
               arr[i] = 8;
         }
         else if (ii == 2)
         {
            arr = new int[] { 0, 0, 1, 1, 2, 2, 3, 3 };
         }
         else if (ii < 6)
         {
             // Lots of zeros
             arr = new int[1<<(ii+6)];

             for (int i=0; i<arr.length; i++)
             {
                 int val = rnd.nextInt(100);

                 if (val >= 33)
                     val = 0;

                 arr[i] = val;
             }                 
         }
         else if (ii == 6)
         {
            // Totally random
            arr = new int[512];

            // Leave zeros at the beginning for ZRLT to succeed
            for (int j=20; j<arr.length; j++)
                arr[j] = rnd.nextInt(range);
         }
         else
         {
            arr = new int[1024];
            // Leave zeros at the beginning for ZRLT to succeed
            int idx = 20;

            while (idx < arr.length)
            {
               int len = rnd.nextInt(40);

               if (len % 3 == 0)
                 len = 1;

                int val = rnd.nextInt(range);
                int end = (idx+len) < arr.length ? idx+len : arr.length;

                for (int j=idx; j<end; j++)
                   arr[j] = val;

               idx += len;
            }
         }

          int size = arr.length;
          ByteFunction f = getByteFunction(name);
          input = new byte[size];
          output = new byte[f.getMaxEncodedLength(size)];
          reverse = new byte[size];
          SliceByteArray sa1 = new SliceByteArray(input, 0);
          SliceByteArray sa2 = new SliceByteArray(output, 0);
          SliceByteArray sa3 = new SliceByteArray(reverse, 0);
          Arrays.fill(output, (byte) 0xAA);

          for (int i=0; i<arr.length; i++)
          {
             input[i] = (byte) (arr[i] & 255);
          }

          System.out.println("\nOriginal: ");

          for (int i=0; i<input.length; i++)
          {
             System.out.print((input[i] & 255) + " ");
          }

          if (f.forward(sa1, sa2) == false)
          {
             // ZRLT may fail if the input data has too few 0s
             if (sa1.index != input.length)
             {
                System.out.println("\nNo compression (ratio > 1.0), skip reverse");
                continue;
             }

             System.out.println("\nEncoding error");
             return false;
          }

          if (sa1.index != input.length)
          {
             System.out.println("\nNo compression (ratio > 1.0), skip reverse");
             continue;
          }

          System.out.println("\nCoded: ");
          //java.util.Arrays.fill(input, (byte) 0);

          for (int i=0; i<sa2.index; i++)
          {
             System.out.print((output[i] & 255) + " "); //+"("+Integer.toBinaryString(output[i] & 255)+") ");
          }

          System.out.println(" (Compression ratio: " + (sa2.index * 100 / input.length)+ "%)");
          f = getByteFunction(name);
          sa2.length = sa2.index;
          sa1.index = 0;
          sa2.index = 0;
          sa3.index = 0;

          if (f.inverse(sa2, sa3) == false)
          {
             System.out.println("Decoding error");
             return false;
          }

          System.out.println("Decoded: ");

          for (int i=0; i<reverse.length; i++)
          {
             System.out.print((reverse[i] & 255) + " ");
          }

          System.out.println();

          for (int i=0; i<input.length; i++)
          {
             if (input[i] != reverse[i])
             {
                System.out.println("Different (index "+i+": "+input[i]+" - "+reverse[i]+")");
                return false;
             }
          }

          System.out.println("Identical");
          System.out.println();
      }
      
      return true;
   }

    
   public static void testSpeed(String name)
   {
      // Test speed
      byte[] input;
      byte[] output;
      byte[] reverse;
      Random rnd = new Random();
      final int iter = name.equals("ROLZ") ? 2000 : 50000;
      final int size = 50000;
      System.out.println("\n\nSpeed test for " + name);
      System.out.println("Iterations: " + iter);
      System.out.println();
      int range = name.equals("ZRLT") ? 5 : 256;

      for (int jj=0; jj<3; jj++)
      {
         ByteFunction f = getByteFunction(name);
         input = new byte[size];
         output = new byte[f.getMaxEncodedLength(size)];
         reverse = new byte[size];
         SliceByteArray sa1 = new SliceByteArray(input, 0);
         SliceByteArray sa2 = new SliceByteArray(output, 0);
         SliceByteArray sa3 = new SliceByteArray(reverse, 0);

         // Generate random data with runs
         // Leave zeros at the beginning for ZRLT to succeed
         int n = iter/20;

         while (n < input.length)        
         {
            byte val = (byte) rnd.nextInt(range);
            input[n++] = val;
            int run = rnd.nextInt(256);
            run -= 220;

            while ((--run > 0) && (n < input.length))       
               input[n++] = val;
         }

         long before, after;
         long delta1 = 0;
         long delta2 = 0;

         for (int ii = 0; ii < iter; ii++)
         {
            f = getByteFunction(name);
            sa1.index = 0;
            sa2.index = 0;
            before = System.nanoTime();

            if (f.forward(sa1, sa2) == false)
            {
               // ZRLT may fail if the input data has too few 0s
               System.out.println("Encoding error");
               continue;
            }

            after = System.nanoTime();
            delta1 += (after - before);
         }

         for (int ii = 0; ii < iter; ii++)
         {
            f = getByteFunction(name);
            sa2.length = sa2.index;
            sa3.index = 0;
            sa2.index = 0;
            before = System.nanoTime();

            if (f.inverse(sa2, sa3) == false)
            {
               System.out.println("Decoding error");
               System.exit(1);
            }

            after = System.nanoTime();
            delta2 += (after - before);
         }

         int idx = -1;

         // Sanity check
         for (int i=0; i<sa1.index; i++)
         {
            if (sa1.array[i] != sa3.array[i])
            {
               idx = i;
               break;
            }
         }

         if (idx >= 0)
            System.out.println("Failure at index "+idx+" ("+sa1.array[idx]+"<->"+sa3.array[idx]+")");
            
         final long prod = (long) iter * (long) size;
         System.out.println(name + " encoding [ms]: " + delta1 / 1000000);
         System.out.println("Throughput [MB/s]: " + prod * 1000000L / delta1 * 1000L / (1024*1024));
         System.out.println(name + " decoding [ms]: " + delta2 / 1000000);
         System.out.println("Throughput [MB/s]: " + prod * 1000000L / delta2 * 1000L / (1024*1024));
      }
   }
}

File: test/TestFunctions.go
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"flag"
	"fmt"
	kanzi "github.com/flanglet/kanzi-go"
	"github.com/flanglet/kanzi-go/function"
	"math/rand"
	"os"
	"strings"
	"time"
)

func main() {
	var name = flag.String("type", "ALL", "Type of function (all, LZ4, ROLZ, SNAPPY, RLT or ZRLT)")

	// Parse
	flag.Parse()
	name_ := strings.ToUpper(*name)
	fmt.Printf("Transform %v", name)

	if name_ == "ALL" {
		fmt.Printf("\n\nTestLZ4")

		if err := TestCorrectness("LZ4"); err != nil {
			os.Exit(1)
		}

		TestSpeed("LZ4")
		fmt.Printf("\n\nTestROLZ")

		if err := TestCorrectness("ROLZ"); err != nil {
			os.Exit(1)
		}

		TestSpeed("ROLZ")
		fmt.Printf("\n\nTestSnappy")

		if err := TestCorrectness("SNAPPY"); err != nil {
			os.Exit(1)
		}

		TestSpeed("SNAPPY")
		fmt.Printf("\n\nTestZRLT")

		if err := TestCorrectness("ZRLT"); err != nil {
			os.Exit(1)
		}

		TestSpeed("ZRLT")
		fmt.Printf("\n\nTestRLT")

		if err := TestCorrectness("RLT"); err != nil {
			os.Exit(1)
		}

		TestSpeed("RLT")
	} else if name_ != "" {
		fmt.Printf("Test%v", name_)

		if err := TestCorrectness(name_); err != nil {
			os.Exit(1)
		}

		TestSpeed(name_)
	}
}

func getByteFunction(name string) (kanzi.ByteFunction, error) {
	switch name {
	case "LZ4":
		res, err := function.NewLZ4Codec()
		return res, err

	case "ROLZ":
		res, err := function.NewROLZCodec(function.ROLZ_LOG_POS_CHECKS)
		return res, err

	case "SNAPPY":
		res, err := function.NewSnappyCodec()
		return res, err

	case "ZRLT":
		res, err := function.NewZRLT()
		return res, err

	case "RLT":
		res, err := function.NewRLT(3)
		return res, err

	default:
		panic(fmt.Errorf("No such byte function: '%s'", name))
	}
}

func TestCorrectness(name string) error {
	fmt.Printf("Correctness test for %v\n", name)
	rng := 256

	if name == "ZRLT" {
		rng = 5
	}

	for ii := 0; ii < 20; ii++ {
		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
		fmt.Printf("\nTest %v\n\n", ii)
		var arr []int

		if ii == 0 {
			arr = []int{0, 1, 2, 2, 2, 2, 7, 9, 9, 16, 16, 16, 1, 3,
				3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3}
		} else if ii == 1 {
			arr = make([]int, 66000)

			for i := range arr {
				arr[i] = 8
			}

			arr[0] = 1
		} else if ii == 2 {
			arr = []int{0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 2, 3, 3, 3}
		} else if ii < 6 {
			// Lots of zeros
			arr = make([]int, 1<<uint(ii+6))

			if rng > 100 {
				rng = 100
			}

			for i := range arr {
				val := rand.Intn(rng)

				if val >= 33 {
					val = 0
				}

				arr[i] = val
			}
		} else if ii == 6 {
			// Totally random
			arr = make([]int, 512)

			// Leave zeros at the beginning for ZRLT to succeed
			for i := 20; i < len(arr); i++ {
				arr[i] = rand.Intn(rng)
			}
		} else {
			arr = make([]int, 1024)
			// Leave zeros at the beginning for ZRLT to succeed
			idx := 20

			for idx < len(arr) {
				length := rnd.Intn(40)

				if length%3 == 0 {
					length = 1
				}

				val := rand.Intn(rng)
				end := idx + length

				if end >= len(arr) {
					end = len(arr) - 1
				}

				for j := idx; j < end; j++ {
					arr[j] = val
				}

				idx += length

			}
		}

		size := len(arr)
		f, err := getByteFunction(name)

		if err != nil {
			fmt.Printf("\nCannot create transform '%v': %v\n", name, err)
			return err
		}

		input := make([]byte, size)
		output := make([]byte, f.MaxEncodedLen(size))
		reverse := make([]byte, size)

		for i := range output {
			output[i] = 0xAA
		}

		for i := range arr {
			input[i] = byte(arr[i])
		}

		f, err = getByteFunction(name)

		if err != nil {
			fmt.Printf("\nCannot create transform '%v': %v\n", name, err)
			return err
		}

		fmt.Printf("\nOriginal: \n")

		for i := range arr {
			fmt.Printf("%v ", input[i])
		}

		srcIdx, dstIdx, err := f.Forward(input, output)

		if err != nil {
			// ZRLT may fail if the input data has too few 0s
			if srcIdx != uint(size) {
				fmt.Printf("\nNo compression (ratio > 1.0), skip reverse")
				continue
			}

			fmt.Printf("\nEncoding error : %v\n", err)
			return err
		}

		if srcIdx != uint(size) {
			fmt.Printf("\nNo compression (ratio > 1.0), skip reverse")
			continue
		}

		fmt.Printf("\nCoded: \n")

		for i := uint(0); i < dstIdx; i++ {
			fmt.Printf("%v ", output[i])
		}

		fmt.Printf(" (Compression ratio: %v%%)\n", int(dstIdx)*100/size)

		f, err = getByteFunction(name)

		if err != nil {
			fmt.Printf("\nCannot create transform '%v': %v\n", name, err)
			return err
		}

		_, _, err = f.Inverse(output[0:dstIdx], reverse)

		if err != nil {
			fmt.Printf("Decoding error : %v\n", err)
			return err
		}

		fmt.Printf("Decoded: \n")

		for i := range reverse {
			fmt.Printf("%v ", reverse[i])
		}

		fmt.Printf("\n")

		// Check
		for i := range reverse {
			if input[i] != reverse[i] {
				fmt.Printf("Different (index %v - %v)\n", input[i], reverse[i])
				return err
			}
		}

		fmt.Printf("Identical\n")
	}

	return error(nil)
}

func TestSpeed(name string) {
	iter := 50000

	if name == "ROLZ" {
		iter = 2000
	}

	size := 50000
	fmt.Printf("\n\nSpeed test for %v\n", name)
	fmt.Printf("Iterations: %v\n", iter)
	rng := 256

	if name == "ZRLT" {
		rng = 5
	}

	for jj := 0; jj < 3; jj++ {
		bf, err2 := getByteFunction(name)

		if err2 != nil {
			fmt.Printf("\nCannot create transform '%v': %v\n", name, err2)
			return
		}

		input := make([]byte, size)
		output := make([]byte, bf.MaxEncodedLen(size))
		reverse := make([]byte, size)
		rand.Seed(int64(jj))

		// Generate random data with runs
		// Leave zeros at the beginning for ZRLT to succeed
		n := iter / 20
		delta1 := int64(0)
		delta2 := int64(0)

		for n < len(input) {
			val := byte(rand.Intn(rng))
			input[n] = val
			n++
			run := rand.Intn(256)
			run -= 220

			for run > 0 && n < len(input) {
				input[n] = val
				n++
				run--
			}
		}

		var err error
		var dstIdx uint

		for ii := 0; ii < iter; ii++ {
			f, err2 := getByteFunction(name)

			if err2 != nil {
				fmt.Printf("\nCannot create transform '%v': %v\n", name, err2)
				return
			}

			before := time.Now()

			_, dstIdx, err = f.Forward(input, output)

			if err != nil {
				fmt.Printf("Encoding error : %v\n", err)
				continue
			}

			after := time.Now()
			delta1 += after.Sub(before).Nanoseconds()
		}

		for ii := 0; ii < iter; ii++ {
			f, err2 := getByteFunction(name)

			if err2 != nil {
				fmt.Printf("\nCannot create transform '%v': %v\n", name, err2)
				return
			}

			before := time.Now()

			if _, _, err = f.Inverse(output[0:dstIdx], reverse); err != nil {
				fmt.Printf("Decoding error : %v\n", err)
				os.Exit(1)
			}

			after := time.Now()
			delta2 += after.Sub(before).Nanoseconds()
		}

		idx := -1

		// Sanity check
		for i := range input {
			if input[i] != reverse[i] {
				idx = i
				break
			}
		}

		if idx >= 0 {
			fmt.Printf("Failure at index %v (%v <-> %v)\n", idx, input[idx], reverse[idx])
			os.Exit(1)
		}

		fmt.Printf("\n%v encoding [ms]: %v", name, delta1/1000000)
		fmt.Printf("\nThroughput [MB/s]: %d", (int64(iter*size))*1000000/delta1*1000/(1024*1024))
		fmt.Printf("\n%v decoding [ms]: %v", name, delta2/1000000)
		fmt.Printf("\nThroughput [MB/s]: %d", (int64(iter*size))*1000000/delta2*1000/(1024*1024))
	}

	println()
}
